core/state_processor.go

```diff
@@ -81,20 +86,12 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 	return receipts, allLogs, *usedGas, nil
 }
 
-// ApplyTransaction attempts to apply a transaction to the given state database
-// and uses the input parameters for its environment. It returns the receipt
-// for the transaction, gas used and an error if the transaction failed,
-// indicating the block was invalid.
-func ApplyTransaction(config *params.ChainConfig, bc ChainContext, author *common.Address, gp *GasPool, statedb *state.StateDB, header *types.Header, tx *types.Transaction, usedGas *uint64, cfg vm.Config) (*types.Receipt, error) {
-	msg, err := tx.AsMessage(types.MakeSigner(config, header.Number))
-	if err != nil {
-		return nil, err
-	}
+func applyTransaction(msg types.Message, config *params.ChainConfig, bc ChainContext, author *common.Address, gp *GasPool, statedb *state.StateDB, header *types.Header, tx *types.Transaction, usedGas *uint64, cfg vm.Config, blockContext vm.BlockContext) (*types.Receipt, error) {
 	// Create a new context to be used in the EVM environment
-	context := NewEVMContext(msg, header, bc, author)
+	txContext := NewEVMTxContext(msg)
 	// Create a new environment which holds all relevant information
 	// about the transaction and calling mechanisms.
-	vmenv := vm.NewEVM(context, statedb, config, cfg)
+	vmenv := vm.NewEVM(blockContext, txContext, statedb, config, cfg)
```

Comment:
Do you really need to instantiate a new evm here? Shouldn't it suffice to do that at block boundaries

Reply:
I don't know, the comment on NewEVM says "The returned EVM is not thread safe and should only ever be used *once*."
I could of course add a _Reset_ method of some kind which would probably greatly improve things.

Reply:
Adding that change resulted in the following test times:
```
BenchmarkInsertChain_empty_memdb-8            	    4536	    262351 ns/op	  185415 B/op	     516 allocs/op
BenchmarkInsertChain_empty_diskdb-8           	    6140	    293717 ns/op	  187502 B/op	     536 allocs/op
BenchmarkInsertChain_valueTx_memdb-8          	    3690	    338649 ns/op	  202926 B/op	     682 allocs/op
BenchmarkInsertChain_valueTx_diskdb-8         	    4546	    324259 ns/op	  206864 B/op	     705 allocs/op
BenchmarkInsertChain_valueTx_100kB_memdb-8    	    1296	   1124334 ns/op	  743574 B/op	     687 allocs/op
BenchmarkInsertChain_valueTx_100kB_diskdb-8   	    1038	   1306500 ns/op	 1059309 B/op	     714 allocs/op
BenchmarkInsertChain_uncles_memdb-8           	    3290	    441751 ns/op	  202184 B/op	     695 allocs/op
BenchmarkInsertChain_uncles_diskdb-8          	    3888	    462218 ns/op	  206863 B/op	     715 allocs/op
BenchmarkInsertChain_ring200_memdb-8          	     189	   6440408 ns/op	 2505123 B/op	   28874 allocs/op
BenchmarkInsertChain_ring200_diskdb-8         	     183	   6861615 ns/op	 2556595 B/op	   28222 allocs/op
BenchmarkInsertChain_ring1000_memdb-8         	     152	   7822861 ns/op	 3117208 B/op	   33133 allocs/op
BenchmarkInsertChain_ring1000_diskdb-8        	     148	   8386961 ns/op	 3207706 B/op	   32458 allocs/op
```
The empty inserts are still a bit slower, but the ring inserts are ~40% faster